lisa/sut_orchestrator/qemu/platform.py

```diff
@@ -0,0 +1,447 @@
+
+        with libvirt.open("qemu:///system") as qemu_conn:
+            try:
+                self._create_nodes(environment, log, qemu_conn)
+                self._fill_nodes_metadata(environment, log, qemu_conn)
+
+            except Exception as ex:
+                self._delete_nodes(environment, log, qemu_conn)
+                raise ex
+
+    # Pre-determine all the nodes' properties, including the name of all the resouces
+    # to be created. This makes it easier to cleanup everything after the test is
+    # finished (or fails).
+    def _configure_nodes(self, environment: Environment, log: Logger) -> None:
+        environment_context = get_environment_context(environment)
+
+        # Generate a random name for the VMs.
+        test_suffix = "".join(random.choice(string.ascii_uppercase) for i in range(5))
+        vm_name_prefix = f"lisa-{test_suffix}"
+
+        environment_context.ssh_public_key = get_public_key_data(
+            self.runbook.admin_private_key_file
+        )
+
+        assert environment.runbook.nodes_requirement
+        for i, node_space in enumerate(environment.runbook.nodes_requirement):
+            assert isinstance(
+                node_space, schema.NodeSpace
+            ), f"actual: {type(node_space)}"
+
+            qemu_node_runbook: QemuNodeSchema = node_space.get_extended_runbook(
+                QemuNodeSchema, type_name=QEMU
+            )
+
+            vm_disks_dir = os.path.dirname(qemu_node_runbook.qcow2)
+
+            node = environment.create_node_from_requirement(node_space)
+            node_context = get_node_context(node)
+
+            node_context.vm_name = f"{vm_name_prefix}-{i}"
+            node_context.cloud_init_file_path = os.path.join(
+                vm_disks_dir, f"{node_context.vm_name}-cloud-init.iso"
+            )
+            node_context.os_disk_base_file_path = qemu_node_runbook.qcow2
+            node_context.os_disk_file_path = os.path.join(
+                vm_disks_dir, f"{node_context.vm_name}-os.qcow2"
+            )
+
+            if not node.name:
+                node.name = node_context.vm_name
+
+            # Read extra cloud-init data.
+            if (
+                qemu_node_runbook.cloud_init
+                and qemu_node_runbook.cloud_init.extra_user_data
+            ):
+                extra_user_data_file_path = str(
+                    constants.RUNBOOK_PATH.joinpath(
+                        qemu_node_runbook.cloud_init.extra_user_data
+                    )
+                )
+                with open(extra_user_data_file_path, "r") as file:
+                    node_context.extra_cloud_init_user_data = yaml.safe_load(file)
+
+    # Create all the VMs.
+    def _create_nodes(
+        self, environment: Environment, log: Logger, qemu_conn: libvirt.virConnect
+    ) -> None:
+        for node in environment.nodes.list():
+            # Create cloud-init ISO file.
+            self._create_node_cloud_init_iso(environment, log, node)
```

Comment:
Does it need to create iso for each node? In theory, LISA supports to deploy different type of nodes in the same environment, but I didn't verify it.

Reply:
The cloud-init ISO sets the hostname to be the same as the VM name to assist with debugging.